In [42]:
import warnings 
warnings.filterwarnings("ignore")

![](../Data/deep.png)

# 패션 mnist

In [43]:
from tensorflow import keras
(train_input, train_target), (test_input, test_target) = keras.datasets.fashion_mnist.load_data()

In [44]:
# train과 검증(valid) 분리
from sklearn.model_selection import train_test_split
train_scaled = train_input / 255 # 시각화 하므로 /255
train_scaled.shape

(60000, 28, 28)

In [45]:
# 2차원으로 변환 
train_scaled = train_scaled.reshape(-1,28*28)
train_scaled.shape

(60000, 784)

In [46]:
train_scaled, val_scaled, train_target, val_target = train_test_split(
    train_scaled,train_target, test_size = 0.2, random_state=42
)

### layer를 추가하는 방법-1 

In [47]:
# layer 만들기
# 은닉층 
# input_shape 입력층
dense1 = keras.layers.Dense(100,activation="sigmoid",input_shape=(784,))
# 출력층
# 다중분류 softmax
dense2= keras.layers.Dense(10,activation="softmax")

In [48]:
# 전체 모델에 층을 추가
model = keras.Sequential([dense1, dense2])

In [49]:
model.summary()
# param # 은 가중치에 절편을 더해진 값
# y = 784 *  100 +10 
# 여기서 784는 가중치 * 100 에 더하기 100한값 => 가중치 갯수에 절편이 더해진 값
# 재실행하면 dense 수 더해짐 

Model: "sequential_7"
_________________________________________________________________
 Layer (type)                Output Shape              Param #   
 dense_11 (Dense)            (None, 100)               78500     
                                                                 
 dense_12 (Dense)            (None, 10)                1010      
                                                                 
Total params: 79,510
Trainable params: 79,510
Non-trainable params: 0
_________________________________________________________________


---
# layer를 추가하는 방법 -2

In [50]:
model = keras.Sequential(
    [
    keras.layers.Dense(100,activation="sigmoid",input_shape=(784,),name='hidden'),
    keras.layers.Dense(10,activation="softmax",name="output")
    ], name = "패션 mnist 모델"
)

In [51]:
model.summary()

Model: "패션 mnist 모델"
_________________________________________________________________
 Layer (type)                Output Shape              Param #   
 hidden (Dense)              (None, 100)               78500     
                                                                 
 output (Dense)              (None, 10)                1010      
                                                                 
Total params: 79,510
Trainable params: 79,510
Non-trainable params: 0
_________________________________________________________________


---
### layer를 추가하는 방법 - 3 

In [52]:
model = keras.Sequential()
model.add(keras.layers.Dense(100,activation="sigmoid",input_shape=(784,)))
model.add(keras.layers.Dense(10,activation="softmax"))

In [53]:
model.summary()

Model: "sequential_8"
_________________________________________________________________
 Layer (type)                Output Shape              Param #   
 dense_13 (Dense)            (None, 100)               78500     
                                                                 
 dense_14 (Dense)            (None, 10)                1010      
                                                                 
Total params: 79,510
Trainable params: 79,510
Non-trainable params: 0
_________________________________________________________________


---
### 모델 훈련 


In [54]:
model.compile(loss="sparse_categorical_crossentropy",metrics = 'accuracy')
model.fit(train_scaled,train_target,epochs=6)

Epoch 1/6
1500/1500 [==============================] - 4s 2ms/step - loss: 0.5611 - accuracy: 0.8080
Epoch 2/6
1500/1500 [==============================] - 3s 2ms/step - loss: 0.4092 - accuracy: 0.8526
Epoch 3/6
1500/1500 [==============================] - 3s 2ms/step - loss: 0.3747 - accuracy: 0.8659
Epoch 4/6
1500/1500 [==============================] - 3s 2ms/step - loss: 0.3516 - accuracy: 0.8723
Epoch 5/6
1500/1500 [==============================] - 3s 2ms/step - loss: 0.3345 - accuracy: 0.8791
Epoch 6/6
1500/1500 [==============================] - 3s 2ms/step - loss: 0.3204 - accuracy: 0.8829


> epochs: 6 => 정확도accuracy: 0.8829 

# sigmoid와 relu 함수의 비교 

### sigmoid function
![](../Data/sigmoid.png)
- sigmoid 함수는 예측값이 0과 1 사이에서 이루어진다
-> 즉 0에도 닿지 않고 1에도 도달하지 않는 값이라 모든 예측값 0에서 1사이이므로 예측값끼리의 차이가 적다. Ex ) 0.3 - 0.1 => 0.2

### relu function
![](../Data/relu.png)
- relu 함수는 예측값을 양수로 만들어 예측값의 차를 크게 만들어준다.
- 여기서 input은 입력층, output은 출력층(예측값)

- sigmoid 함수는 숫자가 무한대로 커지게 되면 데이터의 차이가 거의 없다는 단점이 있다.
- 이를 해결한 것이 relu 함수로 인공신경망에 기여한 바가 크며 음수는 모든 0으로 양수는 항상 최대값을 출력하는 함수

In [57]:
# 입력층을 단순화 하기 
model = keras.Sequential()
# 입력층 
# flatten 입력층을 2차원 그대로 사용한다.
model.add(keras.layers.Flatten(input_shape=(28,28)))
# hidden layer
model.add(keras.layers.Dense(100, activation = "relu"))
model.add(keras.layers.Dense(10, activation = "softmax"))


In [58]:
model.summary()
# output shape의 (행, 열)

Model: "sequential_10"
_________________________________________________________________
 Layer (type)                Output Shape              Param #   
 flatten (Flatten)           (None, 784)               0         
                                                                 
 dense_15 (Dense)            (None, 100)               78500     
                                                                 
 dense_16 (Dense)            (None, 10)                1010      
                                                                 
Total params: 79,510
Trainable params: 79,510
Non-trainable params: 0
_________________________________________________________________


---
### Optimizer (최적화 알고리즘)
- 훈련시 Optimizer를 사용하여 최적화 알고리즘을 구성한다.
- 보통 사용하는 것이 확률적 경사하강법 optimizer이다.

In [ ]:
# adam이 정확도가 높은 편 
# learning rate = 0.01 => sgd
# learning rate = 1 >  => 모멘텀
# learning rate = 0.001(1% 정도) => adam, adagrad
# 전체가 100일 때 

In [60]:
# 적용방법 (확률적 경사 하강법) -1
model.compile(optimizer="sgd",loss='sparse_categorical_crossentropy',metrics='accuracy')

In [61]:
# 적용방법 (확률적 경사 하강법) -2
# instance 
sgd = keras.optimizers.SGD()
model.compile(optimizer=sgd,loss='sparse_categorical_crossentropy',metrics='accuracy')

In [62]:
# 적용방법 (확률적 경사 하강법) , learning rate 변경 -3 : sgd의 learning rate의 기본값:0.01
sgd = keras.optimizers.SGD(learning_rate=0.1)
model.compile(optimizer=sgd,loss='sparse_categorical_crossentropy',metrics='accuracy')

In [67]:
# 적용방법 (mometam)
sgd = keras.optimizers.SGD(momentum=0.9,nesterov =True)
model.compile(optimizer=sgd,loss='sparse_categorical_crossentropy',metrics='accuracy')
# mometam = default 값은 0.9 , 점핑하면서 정답이 아닌 값에서 학습이 끝나지 않게 해 epoch가 줄인다.
# 한번의 jumping 이 epoch 하나

In [68]:
# 적용방법 (adagrad)
adagrad = keras.optimizers.Adagrad()
model.compile(optimizer=adagrad,loss='sparse_categorical_crossentropy',metrics='accuracy')
# mometam = default 값은 0.9

In [69]:
# 적용방법 (RMSprop)
rmsprop = keras.optimizers.RMSprop()
model.compile(optimizer=rmsprop,loss='sparse_categorical_crossentropy',metrics='accuracy')
# mometam = default 값은 0.9